src/viewer.js

```diff
@@ -3548,37 +3572,51 @@ function updateMulti( viewer ) {
     }
 }
 
+function doViewerResize(viewer, containerSize){
+    var viewport = viewer.viewport;
+    var zoom = viewport.getZoom();
+    var center = viewport.getCenter();
+    viewport.resize(containerSize, viewer.preserveImageSizeOnResize);
+    viewport.panTo(center, true);
+    var resizeRatio;
+    if (viewer.preserveImageSizeOnResize) {
+        resizeRatio = THIS[viewer.hash].prevContainerSize.x / containerSize.x;
+    } else {
+        var origin = new $.Point(0, 0);
+        var prevDiag = new $.Point(THIS[viewer.hash].prevContainerSize.x, THIS[viewer.hash].prevContainerSize.y).distanceTo(origin);
+        var newDiag = new $.Point(containerSize.x, containerSize.y).distanceTo(origin);
+        resizeRatio = newDiag / prevDiag * THIS[viewer.hash].prevContainerSize.x / containerSize.x;
+    }
+    viewport.zoomTo(zoom * resizeRatio, null, true);
+    THIS[viewer.hash].prevContainerSize = containerSize;
+    THIS[viewer.hash].forceRedraw = true;
+    THIS[viewer.hash].needsResize = false;
+    THIS[viewer.hash].forceResize = false;
+}
 function updateOnce( viewer ) {
 
     //viewer.profiler.beginUpdate();
 
     if (viewer._opening || !THIS[viewer.hash]) {
         return;
     }
-
-    if (viewer.autoResize) {
-        var containerSize = _getSafeElemSize(viewer.container);
-        var prevContainerSize = THIS[viewer.hash].prevContainerSize;
-        if (!containerSize.equals(prevContainerSize)) {
-            var viewport = viewer.viewport;
-            if (viewer.preserveImageSizeOnResize) {
-                var resizeRatio = prevContainerSize.x / containerSize.x;
-                var zoom = viewport.getZoom() * resizeRatio;
-                var center = viewport.getCenter();
-                viewport.resize(containerSize, false);
-                viewport.zoomTo(zoom, null, true);
-                viewport.panTo(center, true);
-            } else {
-                // maintain image position
-                var oldBounds = viewport.getBounds();
-                viewport.resize(containerSize, true);
-                viewport.fitBoundsWithConstraints(oldBounds, true);
+    if (viewer.autoResize || THIS[viewer.hash].forceResize){
```

Comment:
It doesn't seem like we need `forceResize`... This can just be a check for `needsResize` (along with the `autoResize` check, of course).

Reply:
I thought the point of `forceResize` (as it currently stands, at least) is to allow a developer a way to trigger (force) the resize behavior even if `autoResize==false`. That's why I added a second flag, because otherwise if `autoResize==false` then `needsResize` is ignored.

Reply:
Yes, that's the behavior we want. Now that you mention it, if you take my suggestion and get rid of the `forceResize` flag, you would want to check for `autoResize` in the `ResizeObserver` callback. 

I guess the key difference would be what would happen if you had autoResize off, resized the viewer, then turned it back on. With the separate forceResize flag, it could remember that it got resized and therefore immediately update when autoResize got turned back on. By using needsResize for the forceResize functionality, in that scenario it wouldn't remember that it needed to resize, and therefore it wouldn't resize when you turned autoResize on again. 

I mentioned it because it seemed like a little bit of a pain to have to manage both of those flags, but now I think I've talked myself into wanting them both. 

What do you think? If you agree, then I think we're good here and I can merge it. :)

Reply:
I agree. I think keeping the two flags for managing two different behaviors/usages makes sense. Since they're internal to `Viewer` and used for pretty straightforward purposes, I don't think it should be too painful to keep up with the logic of what's going on for future development. Also, I think your description of the behavior with the two flags - that when `autoResize` is turned on, it should resize immediately if needed - makes good sense for how it should be expected to work.

So, I'm good with moving forward with merging as long as you're satisfied with everything!